voxblox/include/voxblox/mesh/mesh_utils.h

```diff
@@ -0,0 +1,205 @@
+    // We discard triangles where all old vertices were mapped to the same
+    // vertex.
+    size_t new_num_triangle_from_this_block = 0u;
+    for (size_t triangle_idx = 0u; triangle_idx < mesh->indices.size();
+         triangle_idx += 3u) {
+      CHECK_LT(triangle_idx + 2u, mesh->indices.size());
+
+      // Retrieve old vertex indices.
+      size_t vertex_0 = mesh->indices[triangle_idx];
+      size_t vertex_1 = mesh->indices[triangle_idx + 1u];
+      size_t vertex_2 = mesh->indices[triangle_idx + 2u];
+
+      // Make sure the old indices were valid before remapping.
+      CHECK_LT(vertex_0, old_to_new_indices.size());
+      CHECK_LT(vertex_1, old_to_new_indices.size());
+      CHECK_LT(vertex_2, old_to_new_indices.size());
+
+      // Apply vertex index mapping.
+      vertex_0 = old_to_new_indices[vertex_0];
+      vertex_1 = old_to_new_indices[vertex_1];
+      vertex_2 = old_to_new_indices[vertex_2];
+
+      // Make sure the new indices are valid after remapping.
+      CHECK_LT(vertex_0, new_vertex_index);
+      CHECK_LT(vertex_1, new_vertex_index);
+      CHECK_LT(vertex_2, new_vertex_index);
+
+      // Get rid of triangles where all two or three vertices have been
+      // merged.
+      const bool two_or_three_vertex_indices_are_the_same =
+          (vertex_0 == vertex_1) || (vertex_1 == vertex_2) ||
+          (vertex_0 == vertex_2);
+
+      if (!two_or_three_vertex_indices_are_the_same) {
+        connected_mesh->indices.push_back(vertex_0);
+        connected_mesh->indices.push_back(vertex_1);
+        connected_mesh->indices.push_back(vertex_2);
+        ++new_num_triangle_from_this_block;
+      }
+    }
+  }
+
+  // Verify combined mesh.
+  if (connected_mesh->hasColors()) {
+    CHECK_EQ(connected_mesh->vertices.size(), connected_mesh->colors.size());
+  }
+  if (connected_mesh->hasNormals()) {
+    CHECK_EQ(connected_mesh->vertices.size(), connected_mesh->normals.size());
+  }
+
+  // The number of vertices should now be less or equal the amount of triangle
+  // indices, since we merged some of the vertices.
+  // CHECK_LE(connected_mesh->vertices.size(), connected_mesh->indices.size());
+}
+
+inline void createConnectedMesh(
+    const Mesh& mesh, Mesh* connected_mesh,
+    const FloatingPoint approximate_vertex_proximity_threshold = 1e-10) {
+  AlignedVector<Mesh::ConstPtr> meshes;
+  meshes.push_back(Mesh::ConstPtr(&mesh, [](Mesh const*) {}));
+  createConnectedMesh(meshes, connected_mesh,
+                      approximate_vertex_proximity_threshold);
+}
+
+inline void smoothMesh(Mesh* mesh) {
+  for (size_t i = 0; i < mesh->indices.size(); i += 3) {
+    voxblox::Point average_point = (mesh->vertices[mesh->indices[i]] +
+                                    mesh->vertices[mesh->indices[i + 1]] +
+                                    mesh->vertices[mesh->indices[i + 2]]) /
+                                   3.0f;
+    voxblox::Point delta_point =
```

Comment:
could you add a comment explaining what this smoothing function actually does? I'm not sure I fully understand.

Reply:
I'll add a comment. I, didn't end up using this function but it approximates Laplacian smoothing of a mesh (https://en.wikipedia.org/wiki/Laplacian_smoothing) restricted to the normals directions.